src/connectivity/components.c

```diff
@@ -384,39 +390,52 @@ static igraph_error_t igraph_is_connected_weak(const igraph_t *graph, igraph_boo
 igraph_error_t igraph_is_connected(const igraph_t *graph, igraph_bool_t *res,
                         igraph_connectedness_t mode) {
 
+    igraph_cached_property_t prop;
     igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_error_t retval = IGRAPH_SUCCESS;
+    igraph_integer_t no;
 
-    if (no_of_nodes == 0) {
-        /* Changed in igraph 0.9; see https://github.com/igraph/igraph/issues/1539
-         * for the reasoning behind the change */
-        *res = 0;
-        return IGRAPH_SUCCESS;
+    if (!igraph_is_directed(graph)) {
+        mode = IGRAPH_WEAK;
     }
 
-    if (no_of_nodes == 1) {
-        *res = 1;
-        return IGRAPH_SUCCESS;
+    switch (mode) {
+        case IGRAPH_WEAK:
+            prop = IGRAPH_PROP_IS_WEAKLY_CONNECTED;
+            break;
+
+        case IGRAPH_STRONG:
+            prop = IGRAPH_PROP_IS_STRONGLY_CONNECTED;
+            break;
+
+        default:
+            IGRAPH_ERROR("Invalid connectedness mode.", IGRAPH_EINVAL);
     }
 
-    if (mode == IGRAPH_WEAK || !igraph_is_directed(graph)) {
-        return igraph_is_connected_weak(graph, res);
-    } else if (mode == IGRAPH_STRONG) {
-        igraph_error_t retval;
-        igraph_integer_t no;
+    IGRAPH_RETURN_IF_CACHED_BOOL(graph, prop, res);
 
+    if (no_of_nodes == 0) {
+        /* Changed in igraph 0.9; see https://github.com/igraph/igraph/issues/1539
+         * for the reasoning behind the change */
+        *res = 0;
+    } else if (no_of_nodes == 1) {
+        *res = 1;
+    } else if (mode == IGRAPH_WEAK) {
+        retval = igraph_is_connected_weak(graph, res);
+    } else {   /* mode == IGRAPH_STRONG */
         /* A strongly connected graph has at least as many edges as vertices,
          * except for the singleton graph, which is handled above. */
         if (igraph_ecount(graph) < no_of_nodes) {
             *res = 0;
-            return IGRAPH_SUCCESS;
+        } else {
+            retval = igraph_i_connected_components_strong(graph, NULL, NULL, &no);
```

Comment:
Is there any reason you are not using `IGRAPH_CHECK` instead of saving `retval`? Suppose `retval` is an error. Then the value of `*res` will be invalid, but it seems it's still added to the cache.

Reply:
Good point, I have no idea why this was done this way. Just pushed a fix in 676e75ff6